src/exambank/exambank.py

```diff
@@ -0,0 +1,112 @@
+from io import BytesIO
+from pathlib import Path
+
+from pymupdf import Matrix
+from pymupdf import open as p_open
+from pypdf import PdfReader, PdfWriter
+from reportlab.lib import colors
+from reportlab.lib.pagesizes import A4
+from reportlab.lib.units import inch
+from reportlab.pdfgen import canvas
+
+# param to describe margin for exam generation text
+border = 20
+
+def create_watermark(
+        computing_id: str,
+        density: int = 5
+) -> BytesIO:
+    # Generate the tiling watermark
+    stamp_buffer = BytesIO()
+    stamp_pdf = canvas.Canvas(stamp_buffer, pagesize=A4)
+    stamp_pdf.translate(density / 3 * inch, -density / 3 * inch)
+    stamp_pdf.setFillColor(colors.grey, alpha=0.5)
+    stamp_pdf.setFont("Helvetica", 100 / density)
+    stamp_pdf.rotate(45)
+
+    width, height = A4
+
+    for i in range(1, 4  * int(width), int(width/density)):
+        for j in range(1, 4 * int(height), int(height/density)):
+            stamp_pdf.drawCentredString(i, j, computing_id)
+    stamp_pdf.save()
+    stamp_buffer.seek(0)
+
+    # Generate the warning text in corner
+    warning_buffer = BytesIO()
+    warning_pdf = canvas.Canvas(warning_buffer, pagesize=A4)
+    warning_pdf.setFillColor(colors.grey, alpha=0.75)
+    warning_pdf.setFont("Helvetica", 14)
+
+    from datetime import datetime
+    warning_pdf.drawString(border, border, f"This exam was generated by {computing_id} at {datetime.now()}")
+
+    warning_pdf.save()
+    warning_buffer.seek(0)
+
+    # Merge into watermark
+    watermark_pdf = PdfWriter()
+    stamp_pdf = PdfReader(stamp_buffer)
+    warning_pdf = PdfReader(warning_buffer)
+    # Destructively merges in place
+    stamp_pdf.pages[0].merge_page(warning_pdf.pages[0])
+    watermark_pdf.add_page(stamp_pdf.pages[0])
+
+    watermark_buffer = BytesIO()
+    watermark_pdf.write(watermark_buffer)
+    watermark_buffer.seek(0)
+    return watermark_buffer
+
+def apply_watermark(
+        pdf_path: Path | str,
+        # expect a BytesIO instance (at position 0), accept a file/path
```

Comment:
let's not try to accept several different kinds of inputs, but only choose one. This makes the implementation less complicated. Let's just take bytes io. If we want to take a path, let's make a function  `watermark_from_path` or something similar.

Reply:
Sure, resolved in 6ff28fd

Reply:
\<3